src/main.c

```diff
@@ -37,6 +39,71 @@ void print_help(char *argv[]) {
   printf("\tApplication is started when called without argument or after firmware load\n\n");
 }
 
+#include <string.h>
+#define USB_TIMEOUT 5000
+int stlink_dfu_mode(libusb_device_handle *dev_handle, int trigger) {
+    unsigned char data[16];
+    int rw_bytes, res;
+
+    memset(data, 0, sizeof(data));
+
+    data[0] = 0xF9;
+    if (trigger) data[1] = 0x01;
+    /* Write */
+    res = libusb_bulk_transfer(dev_handle,
+                               1 | LIBUSB_ENDPOINT_OUT,
+                               data,
+                               sizeof(data),
+                               &rw_bytes,
+                               USB_TIMEOUT);
+    if (res) {
+        fprintf(stderr, "USB transfer failure\n");
+        return -1;
+    }
+    if (!trigger) {
+        /* Read */
+        libusb_bulk_transfer(dev_handle,
+                             1 | LIBUSB_ENDPOINT_IN,data,
+                             2,
+                             &rw_bytes,
+                             USB_TIMEOUT);
+        if (res) {
+            fprintf(stderr, "stlink_read_infos() failure\n");
+            return -1;
+        }
+    }
+    return data[0] << 8 | data[1];
+}
+
+int test_v21(libusb_context *usb_ctx)
+{
+    libusb_device_handle *dev_handle;
+    int res = 0;
+    dev_handle = libusb_open_device_with_vid_pid(usb_ctx,
+                                                 STLINK_VID,
+                                                 STLINK_PIDV21);
+    if (!dev_handle) {
+        dev_handle = libusb_open_device_with_vid_pid(usb_ctx,
+                                                 STLINK_VID,
+                                                 STLINK_PIDV21_MSD);
+	if (dev_handle)
+	    fprintf(stderr, "StlinkV21_MSD found\n");
+    } else {
+	fprintf(stderr, "StlinkV21 found\n");
+    }
+    if (dev_handle) {
+        res = stlink_dfu_mode(dev_handle, 0);
+        if (res != 0x8000) {
+            libusb_release_interface(dev_handle, 0);
+            return 0;
+        }
+        stlink_dfu_mode(dev_handle, 1);
+        libusb_release_interface(dev_handle, 0);
+    }
+    usleep(1000000);
+    return 1;
+}
```

Comment:
We should ideally this kind of check should be performed when trying to connect to the ST-Link interface. Otherwise this causes redundant connections to the device.